### ISSUE TITLE
Jupyter Collaboration prevents JupyterLab from being killed from the terminal

### ISSUE THREAD

**Issue by martinRenou** [jupyterlab/jupyter_collaboration#161]
# Issue

Installing `jupyter_collaboration` results in not being able to kill JupyterLab from the terminal.

# Reproduce

1. Install `jupyterlab` and `jupyter_collaboration`
2. Create a Notebook, so that you create at least one room
3. Try to kill jupyterlab with `Ctrl + C` in the terminal

# Attempts to fix the issue

After some investigation, it looks like [this `sys.exit` call](https://github.com/jupyterlab/jupyterlab/blob/main/jupyterlab/__main__.py#L8) does not work as expected even though it properly raises a `SystemExit`.

One reason for `sys.exit` to not work can be a hanging thread or subprocess that has not been terminated.

My investigations so far have been:
- Removed [the `_confirm_exit` thread](https://github.com/jupyter-server/jupyter_server/blob/d48843eea78fe2ee37608a767ceea3d0d2f71763/jupyter_server/serverapp.py#L2199) from jupyter_server to see if that thread did not prevent from finishing
- Refactored the tasks cleanup in ypy-websockets and jupyter_collaboration
  - https://github.com/y-crdt/ypy-websocket/pull/76
  - https://github.com/jupyterlab/jupyter_collaboration/pull/160
- Remove any `AnyIO` usage in ypy-websockets. I tried this because there seems to be an "AnyIO worker thread" thread hanging around upon termination. 

None of these attempts fixed the issue. Though I think the refactor of the tasks cleanup could be good to include anyway.

**Comment by rfilmyer** [comment]
Just ran into this again.

`pycrdt-websocket` released [a code change that should have fixed the bug](https://github.com/y-crdt/pycrdt-websocket/pull/148) in 0.16.5, but it doesn't.

> I feel like the "AnyIO worker thread" could be the culprit, but I don't know who creates it.

The answer is the ystore from `jupyter-server-ydoc`.

# Claude's explanation follows:

**What creates it: the ystore.** `sqlite-anyio` routes every single SQLite call
through `anyio.to_thread.run_sync()` (`sqlite_anyio/sqlite.py:56-150`), so
`jupyter-server-ydoc` → `pycrdt-store` → `sqlite-anyio` spawns anyio worker
threads and binds their lifetime to a room task.

**Why it never dies:** anyio's `WorkerThread` is **not a daemon thread**
(`anyio/_backends/_asyncio.py:1044` — `grep -n daemon` on that file returns
nothing), which means that cpython won't finish shutting down until 
`threading._shutdown()` successfully joins it (ie, it returns something). 
`run()` exits only on a `None` sentinel that is delivered
exclusively by `stop()`, registered at line 2678 as:

```python
root_task.add_done_callback(worker.stop, context=Context())
```

So worker shutdown depends entirely on **`root_task` completing**. And
`jupyter-server-ydoc` guarantees it won't, in two places:

`projects/jupyter-server-ydoc/jupyter_server_ydoc/app.py:321` abandons cleanup
after 3s and proceeds with shutdown anyway:

```python
async def stop_extension(self):
    # Cancel tasks and clean up
    await asyncio.wait(
        [
            asyncio.create_task(self.ywebsocket_server.clean()),
            asyncio.create_task(self.file_loaders.clear()),
        ],
        timeout=3,
    )
```

No warning is logged when that timeout fires, so the abandonment is invisible.
And `clean()` itself (`jupyter_server_ydoc/websocketserver.py:60`) has the room
background-task cancellation commented out, behind a FIXME that already
predicts this exact bug:

```python
async def clean(self):
    # TODO: should we wait for any save task?
    self.log.info("Deleting all rooms.")
    # FIXME some clean up should be upstreamed and the following does not
    # prevent hanging stop process - it also requires some thinking about
    # should the ystore write action be cancelled; I guess not as it could
    # results in corrupted data.
    # room_tasks = list()
    # for name, room in list(self.rooms.items()):
    #     for task in room.background_tasks:
    #         task.cancel()  # FIXME should be upstreamed
    #         room_tasks.append(task)
    ...
    await self.stop()
```

Net chain: room background tasks stay pending → the task anyio selected as
`root_task` never completes → the `None` sentinel is never queued → an **idle**
non-daemon worker survives → CPython's `threading._shutdown()` joins it forever
at `Py_Finalize`.

## Important: pycrdt-websocket 0.16.5 is not sufficient

I'm running `pycrdt-websocket` **0.16.5**, which contains
y-crdt/pycrdt-websocket#148 (the `stop()`-waits-for-`start()` fix). I verified
the fix is actually present in my installed copy — `running` property,
`_start_finished` event, and `stop()` awaiting it are all in
`pycrdt/websocket/websocket_server.py`. **The hang still happens.**

That makes sense: #148 makes `WebsocketServer.stop()` wait for its own `start()`
task, but the *room* tasks that anyio bound its worker to are a different set,
and `clean()` explicitly leaves them pending. So this can't be closed by
upgrading pycrdt-websocket; the remaining fix is downstream, here.

## Environment

| | |
|---|---|
| OS | macOS (Darwin 25.6.0), arm64 |
| Python | 3.14.5, GIL enabled (not free-threaded) |
| jupyterlab | 4.6.4 |
| jupyter-server | 2.21.1 |
| jupyter-collaboration | 5.0.4 |
| jupyter-server-ydoc | 3.0.4 |
| pycrdt-websocket | 0.16.5 |
| pycrdt-store | 0.1.5 |
| pycrdt | 0.14.6 |
| sqlite-anyio | 0.2.10 |
| anyio | 4.15.1 |
| tornado | 6.5.10 |

## Log — stops here, never proceeds

```
Shut down this Jupyter server (y/[n])? y
[C 2026-09-23 12:16:41.177 ServerApp] Shutdown confirmed
[I 2026-09-23 12:16:41.190 ServerApp] Shutting down 9 extensions
[I 2026-09-23 12:16:41.227 YDocExtension] Deleting all rooms.
[I 2026-09-23 12:16:41.237 ServerApp] Shutting down 2 kernels
[I 2026-09-23 12:16:41.252 ServerApp] Kernel shutdown: 129a4097-...
[I 2026-09-23 12:16:41.254 ServerApp] Kernel shutdown: ad65f940-...
```

Process stayed alive >12 minutes. The `:8888` LISTEN socket was already gone and
all localhost sockets were `CLOSED`, so it was serving nothing — purely stuck in
finalization.

<details>
<summary>Thread stacks (click to expand)</summary>

`py-spy` is blocked by security policy on this machine, so these are
`/usr/bin/sample` native frames. 2542/2542 samples in both threads — fully
blocked, not spinning.

**Main thread — joining a non-daemon thread during finalization:**

```
Py_Exit
  _Py_Finalize
    wait_for_thread_shutdown            <- threading._shutdown()
      thread_shutdown
        ThreadHandle_join
          _PyParkingLot_Park
            _pthread_cond_wait
              __psynch_cvwait
```

**`AnyIO worker thread` — parked waiting for work that never arrives:**

```
thread_run
  context_run
    method_vectorcall_VARARGS_KEYWORDS
      lock_PyThread_acquire_lock        <- queue.get()
        _PyMutex_LockTimed
          _PyParkingLot_Park
            _pthread_cond_wait
              __psynch_cvwait
```

Worth noting the worker is blocked in `queue.get()`, i.e. **idle between
tasks** — not mid-SQLite-call. `libsqlite3` appears in no live stack, so this is
not SQLite lock contention, and cancelling in-flight ystore writes (the concern
raised in the `clean()` FIXME) is not what's required to fix it. The remaining
threads (`ZMQbg/Reaper`, `ZMQbg/IO/0`) are idle in `kevent`.

</details>

## Minimal reproduction — no Jupyter, no SQLite

This hangs at exit on anyio 4.15.1 / Python 3.14.5:

```python
import asyncio
import threading
import anyio.to_thread

async def pending_forever():
    # first thread-pool use happens in this task -> anyio picks it as root_task
    await anyio.to_thread.run_sync(lambda: None)
    # ...and it never completes, like a room task in flight at shutdown
    await asyncio.Event().wait()

def main():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    task = loop.create_task(pending_forever())
    loop.call_later(0.5, loop.stop)   # tornado: IOLoop.start() == run_forever()
    loop.run_forever()
    print(f"root task done={task.done()}")
    for t in threading.enumerate():
        if t is not threading.main_thread():
            print(f"  {t.name!r} daemon={t.daemon}")

main()
```

```
root task done=False
  'AnyIO worker thread' daemon=False
<hangs forever>
```

The same script driven by `asyncio.run()` instead of `run_forever()` exits 0,
because the real `run_until_complete` root task completes and the sentinel gets
delivered. I've filed the upstream design question separately on the anyio
tracker; this comment is about the downstream half.

## Suggested fix here

1. Un-comment and finish the room-task cancellation in `clean()`. Per the stacks
   above, the worker is idle, so this does **not** require cancelling in-flight
   ystore writes — the FIXME's data-corruption worry doesn't block the fix.
2. Don't silently abandon `clean()` when the 3s `asyncio.wait` timeout fires in
   `stop_extension()` — at minimum log which tasks were left pending, so this
   failure mode is visible in the log instead of being a silent hang.

## Caveat

About a minute before this shutdown, I deleted one notebook's room rows from
`.jupyter_ystore.db` with the `sqlite3` CLI while the server was still running.
I don't believe it's related — the block is a Python `threading.Lock` in
`queue.get()` with no `libsqlite3` frames anywhere, and the minimal repro above
reproduces the hang with no SQLite involved at all — but disclosing it in case
it matters.
